csharp/src/Ice/Communicator-OutgoingConnectionFactory.cs

```diff
@@ -0,0 +1,270 @@
+                        }
+                    }
+                    // Set the callback used to remove the connection from the factory.
+                    connection.Remove = connection => Remove(connection);
+                    return connection;
+                }
+                catch (TransportException)
+                {
+                    _transportFailures[endpoint] = DateTime.Now;
+                    throw;
+                }
+                finally
+                {
+                    lock (_mutex)
+                    {
+                        // Don't modify the pending connections map after the communicator was disposed.
+                        if (_disposeTask == null)
+                        {
+                            _pendingOutgoingConnections.Remove((endpoint, label));
+                        }
+                    }
+                }
+            }
+        }
+
+        internal Connection? GetConnection(List<Endpoint> endpoints, NonSecure preferNonSecure, object? label)
+        {
+            lock (_mutex)
+            {
+                foreach (Endpoint endpoint in endpoints)
+                {
+                    if (_outgoingConnections.TryGetValue((endpoint, label),
+                                                         out LinkedList<Connection>? connections) &&
+                        connections.FirstOrDefault(
+                            connection => connection.IsActive && connection.CanTrust(preferNonSecure))
+                        is Connection connection)
+                    {
+                        return connection;
+                    }
+                }
+                return null;
+            }
+        }
+
+        internal List<Endpoint> OrderEndpointsByTransportFailures(List<Endpoint> endpoints)
+        {
+            if (_transportFailures.IsEmpty)
+            {
+                return endpoints;
+            }
+            else
+            {
+                // Purge expired transport failures
+
+                // TODO avoid purge failures with each call
+                DateTime expirationDate = DateTime.Now - TimeSpan.FromSeconds(5);
+                foreach ((Endpoint endpoint, DateTime date) in _transportFailures)
+                {
+                    if (date <= expirationDate)
+                    {
+                        _ = ((ICollection<KeyValuePair<Endpoint, DateTime>>)_transportFailures).Remove(
+                            new KeyValuePair<Endpoint, DateTime>(endpoint, date));
+                    }
+                }
+
+                return endpoints.OrderBy(
+                    endpoint => _transportFailures.TryGetValue(endpoint, out DateTime value) ? value : default).ToList();
+            }
+        }
+
+        internal void Remove(Connection connection)
+        {
+            lock (_mutex)
+            {
+                LinkedList<Connection> list = _outgoingConnections[(connection.Endpoint, connection.Label)];
```

Comment:
Shouldn't we check if the communicator is disposed here? Doesn't the communicator dispose async assumes _outgoingConnecitons is immutable once _disposeTask is set?

Reply:
I think it is safe because we hold the mutex when creating the GoAwayAsync tasks, so there is no problem with _outgoingConnections map being modified. The lock is only released when we await WhenAll. https://github.com/zeroc-ice/ice/blob/5d77f9fb63caa29e90bbbc392215cc70d87a0c4e/csharp/src/Ice/Communicator.cs#L919